crates/dbsp/src/tutorial.rs

```diff
@@ -1950,6 +1951,298 @@
+//!                     })
+//!                     .join(
+//!                         &edges
+//!                             .map_index(|Tup3(from, to, weight)| (*from, Tup3(*from, *to, *weight))),
+//!                         |_end_from,
+//!                          Tup4(start, _end, cum_weight, hopcnt),
+//!                          Tup3(_from, to, weight)| {
+//!                             Tup4(*start, *to, cum_weight + weight, hopcnt + 1)
+//!                         },
+//!                     )
+//!                     // You can think of the `plus` operator to something
+//!                     // similar to the `union` operator in SQL.
+//!                     .plus(&len_1);
+//!
+//!                 Ok(len_n)
+//!             },
+//!         )?;
+//!
+//!         let mut expected_outputs = ([
+//!             // We expect the full transitive closure in the first step.
+//!             zset! {
+//!                 Tup4(0, 1, 1, 1) => 1,
+//!                 Tup4(0, 2, 2, 2) => 1,
+//!                 Tup4(0, 3, 4, 3) => 1,
+//!                 Tup4(0, 4, 6, 4) => 1,
+//!                 Tup4(1, 2, 1, 1) => 1,
+//!                 Tup4(1, 3, 3, 2) => 1,
+//!                 Tup4(1, 4, 5, 3) => 1,
+//!                 Tup4(2, 3, 2, 1) => 1,
+//!                 Tup4(2, 4, 4, 2) => 1,
+//!                 Tup4(3, 4, 2, 1) => 1,
+//!             },
+//!             // These paths are removed in the second step.
+//!             zset! {
+//!                 Tup4(0, 2, 2, 2) => -1,
+//!                 Tup4(0, 3, 4, 3) => -1,
+//!                 Tup4(0, 4, 6, 4) => -1,
+//!                 Tup4(1, 2, 1, 1) => -1,
+//!                 Tup4(1, 3, 3, 2) => -1,
+//!                 Tup4(1, 4, 5, 3) => -1,
+//!             },
+//!         ] as [_; STEPS])
+//!             .into_iter();
+//!
+//!         closure.inspect(move |output| {
+//!             assert_eq!(*output, expected_outputs.next().unwrap());
+//!         });
+//!
+//!         Ok(closure.output())
+//!     })?;
+//!
+//!     for i in 0..STEPS {
+//!         let iteration = i + 1;
+//!         println!("Iteration {} starts...", iteration);
+//!         circuit_handle.step()?;
+//!         output_handle.consolidate().iter().for_each(
+//!             |(Tup4(start, end, cum_weight, hopcnt), _, z_weight)| {
+//!                 println!(
+//!                     "{start} -> {end} (cum weight: {cum_weight}, hops: {hopcnt}) => {z_weight}"
+//!                 );
+//!             },
+//!         );
+//!         println!("Iteration {} finished.", iteration);
+//!     }
+//!
+//!     Ok(())
+//! }
+//! ```
+//!
+//! Finally, we point out that introducing a cycle to the graph prevents
+//! this fixed-point computation to stop terminating because then
+//! there is no fixed-point anymore. So pay attention to your data and
+//! your queries, if using more powerful (but somewhat more dangerous) iterative
+//! queries.
```

Comment:
```suggestion
//! there is no fixed-point anymore.
```

I don't think that paying attention to data is the lesson here. The fix explained below is :)